distributed/protocol/core.py

```diff
@@ -87,19 +138,50 @@ def _decode_default(obj):
                     frames[offset],
                     object_hook=msgpack_decode_default,
                     use_list=False,
-                    **msgpack_opts
+                    **msgpack_opts,
                 )
                 offset += 1
                 sub_frames = frames[offset : offset + sub_header["num-sub-frames"]]
+                if "callable" in sub_header:
+                    if deserialize:
+                        try:
+                            return loads_function(sub_header["callable"])
+                        except Exception as e:
+                            if deserialize == "delay-exception":
+                                return DelayedExceptionRaise(e)
+                            else:
+                                raise
+                    else:
+                        return SerializedCallable(sub_header, sub_frames)
                 if deserialize:
                     if "compression" in sub_header:
                         sub_frames = decompress(sub_header, sub_frames)
-                    return merge_and_deserialize(
-                        sub_header, sub_frames, deserializers=deserializers
-                    )
+                    try:
+                        return merge_and_deserialize(
+                            sub_header, sub_frames, deserializers=deserializers
+                        )
+                    except Exception as e:
+                        if deserialize == "delay-exception":
+                            return DelayedExceptionRaise(e)
+                        else:
+                            raise
                 else:
                     return Serialized(sub_header, sub_frames)
             else:
+                # Notice, even though `msgpack_decode_default()` supports
+                # `__MsgpackList__`, we decode it here explicitly. This way
+                # we can delay the convertion to a regular `list` until it
+                # gets to a worker.
+                if "__MsgpackList__" in obj:
```

Comment:
What is the type of obj here?  Is `in` the right test here, or is this special value in a more specific place?